### PR TITLE
fips: setup the FIPS provider in pendantic mode for testing

### PR DESCRIPTION
**Pull request by paulidale**
Add options so that the FIPS provider is set to pedantic mode for testing purposes.

Kind of related to and branched from #20752 but not related to installation.

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
OTC: **should this be merged into the 3.1 branch?**

File: util/mk-fipsmodule-cnf.pl
Comment (t8m) [inline util/mk-fipsmodule-cnf.pl:16]: I'm not sure these default values are right as this creates the fipsmodule.cnf to be used by make install.

Comment (paulidale) [reply util/mk-fipsmodule-cnf.pl]: Fixed, along with the test that used the wrong one.

**Comment by paulidale** [comment]
OTC: this is a bug fix (not testing FIPS mode) not a feature.

**Comment by paulidale** [comment]
Ping @openssl/committers for first review.

File: apps/fipsinstall.c
Comment (slontis) [inline apps/fipsinstall.c:99]: Maybe leave the last ',' out on these tables to avoid accidents when adding additional options..

File: apps/fipsinstall.c
Comment (slontis) [inline apps/fipsinstall.c:399]: Dont like this copy here much.. Why not have a variable that points to constant tables?

Comment (paulidale) [reply apps/fipsinstall.c:399]: Because the table needs to be non-const unless this option is set.  Other options alter it's contents.

The structure occupies one byte currently (probably rounds up to four), the copy isn't an overhead.

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:99]: I left it in to reduce clutter in the diffs when adding additional options…

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:399]: I could `memset(fips_opts, 255, sizeof(fips_opts))` which avoids the copy but relies on all FIPS options being set bits/flags.

Comment (slontis) [reply apps/fipsinstall.c:399]: Non leave as it is - I missed the const on the first table but not the second.

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:99]: Given that fields can be added anywhere, I'm not sure how leaving the comma off helps.

File: test/ssl-tests/30-extended-master-secret.cnf.in
Comment (slontis) [inline test/ssl-tests/30-extended-master-secret.cnf.in:30]: What is this version check achieving currently? It seems to skip the test if it doesnt match.. (Wont it always match with <=3.1)

File: apps/fipsinstall.c
Comment (slontis) [reply apps/fipsinstall.c:99]: Doesnt help with the order, only the size if you forget to add one?

File: test/ssl-tests/30-extended-master-secret.cnf.in
Comment (paulidale) [reply test/ssl-tests/30-extended-master-secret.cnf.in:30]: Yes, skip if no match.

- 3.1.0 doesn't have mandatory EMS support, so it can run this test.
- 3.1.1 (& later) will have mandatory EMS support, so they can't run it.

Comment (slontis) [reply test/ssl-tests/30-extended-master-secret.cnf.in:30]: Maybe put this info back into the commit message?

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:99]: It won't help there.  Extra fields will be zero filled regardless of the comma or not.
C doesn't have a way to mandate initialisation of all fields.

File: test/ssl-tests/30-extended-master-secret.cnf.in
Comment (paulidale) [reply test/ssl-tests/30-extended-master-secret.cnf.in:30]: Added.

**Comment by slontis** [comment]
Does this need to be rebased after the other PR is merged?

**Comment by slontis** [comment]
i.e it is dependant on the other PR being merged first?

**Review by slontis — APPROVED** [review]
Approved subject to tests passing

**Comment by paulidale** [comment]
Current the rebase will change nothing.  This is current with the other one.

**Comment by paulidale** [comment]
@openssl/committers still pending second review

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.1. Thank you.

**Pull request closed without merging**
